### PR TITLE
Support replicated table in GPDB

### PR DESCRIPTION
**Pull request by pengzhout**
Support replicated table in GPDB

    Currently, tables are distributed across all segments by hash or random in GPDB. There
    are requirements to introduce a new table type that all segments have the duplicate
    and full table data called replicated table.

    To implement it, we added a new distribution policy named POLICYTYPE_REPLICATED to mark
    a replicated table and added a new locus type named CdbLocusType_SegmentGeneral to specify
    the distribution of tuples of a replicated table.  CdbLocusType_SegmentGeneral implies
    data is generally available on all segments but not available on qDisp, so plan node with
    this locus type can be flexibly planned to execute on either single QE or all QEs. it is
    similar with CdbLocusType_General, the only difference is that CdbLocusType_SegmentGeneral
    node can't be executed on qDisp. To guarantee this, we try our best to add a gather motion
    on the top of a CdbLocusType_SegmentGeneral node when planing motion for join even other
    rel has bottleneck locus type, a problem is such motion may be redundant if the single QE
    is not promoted to executed on qDisp finally, so we need to detect such case and omit the
    redundant motion at the end of apply_motion(). We don't reuse CdbLocusType_Replicated since
    it's always implies a broadcast motion bellow, it's not easy to plan such node as direct
    dispatch to avoid getting duplicate data.

    We don't support replicated table with inherit/partition by clause now, the main problem is
    that update/delete on multiple result relations can't work correctly now, we can fix this
    later.

### CODE REVIEW

**Comment by adam8157** [comment]
Great work!

Better to rebase on the latest master, I saw some duplicated changes, auto-merging might cause some issues for this big PR.

`COPY ON SEGMENT` is not handled, I suggest you disable COPY FROM/TO replicated tables, at least for now.

File: src/backend/cdb/cdbcat.c [hunk 1]
```diff
@@ -209,13 +215,20 @@ GpPolicyFetch(MemoryContext mcxt, Oid tbloid)
...
-		policy->nattrs = nattrs;
+		policy = makeGpPolicy(mcxt, POLICYTYPE_PARTITIONED, nattrs);
+
 		for (i = 0; i < nattrs; i++)
 		{
 			policy->attrs[i] = attrnums[i];
 		}
+
+		if (nattrs == 1 &&
+			policy->attrs[0] == GpReplicatedTableAttributerNumberIdentifier)
```
Comment (hlinnaka) [inline src/backend/cdb/cdbcat.c]: This magic attribute number to mean "replicated" is quite a hack. Let's just add a new "policytype" column to the gp_policy table.

File: src/backend/cdb/cdbpath.c [hunk 2]
```diff
@@ -134,6 +134,33 @@ cdbpath_create_motion_path(PlannerInfo *root,
 			return (Path *) pathnode;
 		}
 
+		if (CdbPathLocus_IsSegmentGeneral(subpath->locus))
+		{
+			/*
+			 * Data is only available on segments, to distingush it with
```
Comment (hlinnaka) [inline src/backend/cdb/cdbpath.c]: I don't understand this comment.

File: src/backend/cdb/cdbpathlocus.c [hunk 3]
```diff
@@ -402,7 +406,16 @@ cdbpathlocus_from_subquery(struct PlannerInfo *root,
 			if (flow->segindex == -1)
 				CdbPathLocus_MakeEntry(&locus);
 			else
-				CdbPathLocus_MakeSingleQE(&locus);
+			{
+				/*
+				 * keep segmentGeneral character, otherwise planner may put
```
Comment (hlinnaka) [inline src/backend/cdb/cdbpathlocus.c]: Why is this needed? Got an example query? Seems wrong to make the locus SegmentGeneral, if singleton flow was requested.

File: src/backend/executor/execMain.c [hunk 4]
```diff
@@ -158,6 +158,31 @@ void ExecCheckRTEPerms(RangeTblEntry *rte);
...
+static void AdjustReplicatedTableCounts(EState *estate);
+
+/*
+ * Support for SELECT INTO (a/k/a CREATE TABLE AS)
+ *
+ * We implement SELECT INTO by diverting SELECT's normal output with
+ * a specialized DestReceiver type.
+ */
+
+typedef struct
```
Comment (hlinnaka) [inline src/backend/executor/execMain.c]: Moving the struct is not cool, because we got it from the upstream. This will create merge conflicts as we merge with upstream.

File: src/backend/executor/execUtils.c [hunk 5]
```diff
@@ -1813,6 +1815,58 @@ InventorySliceTree(Slice ** sliceMap, int sliceIndex, SliceReq * req)
 	}
 }
 
+/*
+ * Helper function to adjust slice tree for replicated table.
+ *
+ * Main slice or init plans that working on replicated table may
```
Comment (hlinnaka) [inline src/backend/executor/execUtils.c]: I don't understand this. Why do we ever create 1-size gangs to begin with, if we convert all of them to N-size gangs at the end? And why do you need N-size gangs, isn't it enough to read from one segment, if the table is replicated?

File: src/backend/parser/gram.y [hunk 6]
```diff
@@ -3941,6 +3941,7 @@ OptConsTableSpace:   USING INDEX TABLESPACE name	{ $$ = $4; }
 
 DistributedBy:   DISTRIBUTED BY  '(' columnListUnique ')'		{ $$ = $4; }
 			| DISTRIBUTED RANDOMLY			{ $$ = list_make1(NULL); }
+			| DISTRIBUTED FULLY			{ $$ = list_make2(NULL, NULL); }
```
Comment (hlinnaka) [inline src/backend/parser/gram.y]: I think we need a better representation of RANDOMLY and FULLY. Using a one-element list with NULL element for RANDOMLY was a hack to begin with, using a two-element list with two NULLS for FULLY is even more so.

File: src/backend/executor/execUtils.c (hunk 5 above)
Comment (d) [reply src/backend/executor/execUtils.c]: Isn't that a cost-based decision? What if you're doing a nest-loop join between a distributed table `foo` and an "every-segment" table `bar`, wouldn't it be often beneficial to read the same  `bar` from every segment?

Comment (pengzhout) [reply src/backend/executor/execUtils.c]: @d we do read the same bar from every segment and there is no motion added if 'every-segment' join with partition table, there are rules defined in the cdbpath_motion_for_join(), please refer to that function.

**Comment by danielgustafsson** [comment]
I don't see a check for prohibiting exchanging a replicated table into a partitioning hierarchy via `ALTER TABLE .. EXCHANGE ..`, is that intentional?

File: src/backend/executor/execUtils.c (hunk 5 above)
Comment (pengzhout) [reply src/backend/executor/execUtils.c]: @hlinnaka you are right, it just need to read from one segment, but there is a limitation in GPDB: On segments, readers/writers need to get same snapshot, so readers always need to wait writer creating a snapshot and reuse it. However, only N-size gang on segments can possibly act as writer gang, others is reader, so in our case we need to promote 1-size gang to N-size gang as needed to bypass this limitation.

File: src/backend/executor/execMain.c (hunk 4 above)
Comment (pengzhout) [reply src/backend/executor/execMain.c]: will fix this, thanks.

**Comment by hlinnaka** [comment]
On 12/02/18 10:20, Tang Pengzhou wrote:
> pengzhout commented on this pull request.
> 
> 
> 
>> @@ -1813,6 +1815,58 @@ InventorySliceTree(Slice ** sliceMap, int sliceIndex, SliceReq * req)
>   	}
>   }
>   
> +/*
> + * Helper function to adjust slice tree for replicated table.
> + *
> + * Main slice or init plans that working on replicated table may
> 
> @hlinnaka you are right, it just need to read from one segment, but
> there is a limitation in GPDB: On segments, readers/writers need to
> get same snapshot, so readers always need to wait writer creating a
> snapshot and reuse it. However, only N-size gang on segments can
> possibly act as writer gang, others is reader, so in our case we need
> to promote 1-size gang to N-size gang as needed to bypass this
> limitation.

Oh, that's subtle. Can't a 1-size reader gang reuse the snapshot created 
by a N-sized writer gang? Can we lift that limitation?

- Heikki

File: src/backend/cdb/cdbpathlocus.c (hunk 3 above)
Comment (pengzhout) [reply src/backend/cdb/cdbpathlocus.c]: For example:  t1 :  partition table, t2: replicated table.
and query is:
select * from t1, (select * from t2 order by c1) as t2;

if planner set the locus of "select * from t2 order by c1" to singleQE, planner may add a broadcast node on the subqueryscan node when joining with t1, it should be wrong.

BTW, SegmentGeneral node is assigned a singleton by default.

File: src/backend/cdb/cdbpath.c (hunk 2 above)
Comment (pengzhout) [reply src/backend/cdb/cdbpath.c]: This is what i said to guarantee node with segmentGeneral always executed on segment, normally no motion needed if you want move a segmentGeneral path to singleQE path, but if we
explicit add a gather motion here, segmentGeneral node may be promoted to execute on QD.

For example:
t1: partition table  t2: replicated table
query is : select * from t2, (select * from t1 limit 1) as sub_t;

if we didn't add a gather motion, the plan will looks like.

Nestloop
   ---> scan on t2
   --->  limit   
        ---> gather motion (3:1)
                --> scan on t1

obviously, it is a wrong plan, if we explicated add a motion, the plan will looks like:
Nestloop
   ---> gather motion (1:1)
        --> scan on t2
   --->  limit   
        ---> gather motion (3:1)
                --> scan on t1

in apply_motion_mutator(), it will recheck if the motion is redundant and omit it if needed.

File: src/backend/cdb/cdbcat.c (hunk 1 above)
Comment (pengzhout) [reply src/backend/cdb/cdbcat.c]: Thanks, I also realize the magic number and the two NULL list thing within parser are too hack, if we can modify the catalog at this stage, I'm totally OK to abandon this hack way.

**Comment by pengzhout** [comment]
@hlinnaka Yes, they can reuse, but if plan has only one slice and the only slice is a 1-size gang, there is no way but promoting the 1-size to N-size.

**Comment by pengzhout** [comment]
@adam8157 Thanks, maybe we can have a discussion offline for the copy support thing.

**Comment by pengzhout** [comment]
@danielgustafsson Thanks, It should be a missing check, I will take a look.

**Comment by ashwinstar** [comment]
I didn't read the code much but just played with it little bit and loved it ! Great work !

Just curious on these points:
- Inserting `random()` values worked perfectly, would like to seek knowledge how this works. Does master evaluate the `random()` and push the values to every segment ?
```
CREATE TABLE a (a float, b float) distributed fully;
INSERT into a SELECT 1,random() FROM generate_series(1,10);
```

- ```SELECT gp_segment_id, * FROM a``` always returned from segment 2. So, for simple selects does it randomly pick one segment or hard-coded to pull from a segment ?

- On expansion (gpexpand), how will the data be replicated to newly added nodes ?

**Comment by pengzhout** [comment]
@ashwinstar Thanks a lot
1. Seems random() is just evaluated on a single QE and been broadcasted to each segments.
2. For gp_segment_id thing, we use 'gp_session_id % numSegments' to decide the single segments. 
3. gpexpand, ORCA and other stuffs are still in the to-do-list now, so frankly, I have no idea how gpexpand will be implemented now.

File: src/backend/cdb/cdbmutate.c [hunk 7]
```diff
@@ -436,24 +436,54 @@ apply_motion(PlannerInfo *root, Plan *plan, Query *query)
...
 
-				Assert(query->intoPolicy->ptype == POLICYTYPE_PARTITIONED);
+				if (query->intoPolicy->ptype == POLICYTYPE_REPLICATED)
+				{
+					/*
+					 * CdbLocusType_SegmentGeneral is only used by replicated table right now,
+					 * so if both input and target are replicated table, no need to add a motion
+					 */
+					if (plan->flow->flotype == FLOW_SINGLETON &&
+							plan->flow->locustype == CdbLocusType_SegmentGeneral)
```
Comment (pengzhout) [inline src/backend/cdb/cdbmutate.c]: @ashwinstar just reminds me that it maybe not right to do the optimization here if plan contain volatile functions. need a further filter here or just abandon this optimization.

File: src/backend/parser/gram.y (hunk 6 above)
Comment (vraghavan78) [reply src/backend/parser/gram.y]: Agree with @hlinnaka. Here is a suggestion. In Orca, we define:

* UNIVERSAL --- Data available every where including master (eg. ValueScan of constants)
* REPLICATED --- Data available on the segments and not on the master. 

So I would suggest a similar proposal. Why not say `DISTRIBUTED REPLICATED` and `DISTRIBUTED UNIVERSALLY`

**Comment by ashwinstar** [comment]
On Wed, Feb 14, 2018 at 10:42 AM, Venkatesh Raghavan <
notifications@github.com> wrote:

> *@vraghavan78* commented on this pull request.
> ------------------------------
>
> In src/backend/parser/gram.y
> <https://github.com/greenplum-db/gpdb/pull/4536#discussion_r168271158>:
>
> > @@ -3941,6 +3941,7 @@ OptConsTableSpace:   USING INDEX TABLESPACE name	{ $$ = $4; }
>
>  DistributedBy:   DISTRIBUTED BY  '(' columnListUnique ')'		{ $$ = $4; }
>  			| DISTRIBUTED RANDOMLY			{ $$ = list_make1(NULL); }
> +			| DISTRIBUTED FULLY			{ $$ = list_make2(NULL, NULL); }
>
> Agree with @hlinnaka <https://github.com/hlinnaka>. Here is a suggestion.
> In Orca, we define:
>
>    - UNIVERSAL --- Data available every where including master (eg.
>    ValueScan of constants)
>    - REPLICATED --- Data available on the segments and not on the master.
>
> So I would suggest a similar proposal. Why not say DISTRIBUTED REPLICATED
> and DISTRIBUTED UNIVERSALLY
>
DISTRIBUTED UNIVERSALLY sounds nice :-)

**Comment by bhuvnesh2703** [comment]
Noting down feedback/issues as we see them.
1. \h help does not show the fully (or whatever name chose) option for replicate tables
```
  [ DISTRIBUTED BY (column, [ ... ] ) | DISTRIBUTED RANDOMLY ]
```

2. With ORCA On, the below CTAS fails with this branch.
```
create table t6 (a int, b int ) distributed randomly;
pivotal=# create table t7 as select * from t6;
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause. Creating a NULL policy entry.
ERROR:  unrecognized node type: 2139062143 (copyfuncs.c:5635)
```
Had to add in this patch in ORCA translator to fix it
```
[gpdb: pivotal{rpt_demo_pr_v2}]$ git diff
diff --git a/src/backend/gpopt/gpdbwrappers.cpp b/src/backend/gpopt/gpdbwrappers.cpp
index 7533045..41a99f7 100644
--- a/src/backend/gpopt/gpdbwrappers.cpp
+++ b/src/backend/gpopt/gpdbwrappers.cpp
@@ -3151,4 +3151,16 @@ gpdb::FAbortRequested
 {
        return (QueryCancelPending || ProcDiePending);
 }
+
+GpPolicy *
+gpdb::PMakeGpPolicy
+       (
+               MemoryContext mcxt,
+               GpPolicyType ptype,
+               int nattrs
+       )
+{
+       return makeGpPolicy(mcxt, ptype, nattrs);
+}
 // EOF
+
diff --git a/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp b/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp
index 7305b6f..ac6eaa2 100644
--- a/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp
+++ b/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp
@@ -5111,9 +5111,11 @@ CTranslatorDXLToPlStmt::PdistrpolicyFromCtas
        {
                ulDistrColsAlloc = ulDistrCols;
        }
+
+       GpPolicy *pdistrpolicy = gpdb::PMakeGpPolicy(NULL, POLICYTYPE_PARTITIONED, ulDistrColsAlloc);

-       GpPolicy *pdistrpolicy = (GpPolicy *) gpdb::GPDBAlloc(sizeof(GpPolicy) +
-                                                                                                                       ulDistrColsAlloc * sizeof(AttrNumber));
+       //GpPolicy *pdistrpolicy = (GpPolicy *) gpdb::GPDBAlloc(sizeof(GpPolicy) +
+       //                                                                                                              ulDistrColsAlloc * sizeof(AttrNumber));
        GPOS_ASSERT(IMDRelation::EreldistrHash == pdxlop->Ereldistrpolicy() ||
                                IMDRelation::EreldistrRandom == pdxlop->Ereldistrpolicy());

diff --git a/src/include/gpopt/gpdbwrappers.h b/src/include/gpopt/gpdbwrappers.h
index 53f80ca..6750865 100644
--- a/src/include/gpopt/gpdbwrappers.h
+++ b/src/include/gpopt/gpdbwrappers.h
@@ -646,6 +646,8 @@ namespace gpdb {
        // returns true if a query cancel is requested in GPDB
        bool FAbortRequested(void);

+       GpPolicy *PMakeGpPolicy(MemoryContext mcxt, GpPolicyType ptype, int nattrs);
+
```
Table was created after patch
```
pivotal=# create table t8 as select * from t7;
LOG:  statement: create table t9 as select * from t7;
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause. Creating a NULL policy entry.
SELECT 0
```

BC & SD

File: src/backend/parser/parse_utilcmd.c [hunk 8]
```diff
@@ -1685,7 +1707,19 @@ transformDistributedBy(ParseState *pstate, CreateStmtContext *cxt,
...
+
+		if (distributedBy->length == 1 && linitial(distributedBy) == NULL)
+		{
+			/* do nothing */
+		}
+		else if (distributedBy->length == 2 &&
+			linitial(distributedBy) == NULL &&
+			lsecond(distributedBy) == NULL)
+		{
+			policy->ptype = POLICYTYPE_REPLICATED;
```
Comment (bhuvnesh2703) [inline src/backend/parser/parse_utilcmd.c]: may be just add a comment that while parsing distributed fully the list contains `list_make2(NULL, NULL)`; so `length = 2`

File: src/backend/parser/parse_utilcmd.c [hunk 9]
```diff
@@ -3566,7 +3606,8 @@ getLikeDistributionPolicy(InhRelation* e)
 			}
 		}
 		else
-		{	/* old table is distributed randomly. */
+		{
+			/* old table is distributed fully. */
```
Comment (bhuvnesh2703) [inline src/backend/parser/parse_utilcmd.c]: the comment appears wrong, should this be distributed randomly as the list contains 1 NULL element.?

File: src/bin/psql/describe.c [hunk 10]
```diff
@@ -2716,7 +2716,11 @@ add_distributed_by_footer(const char* oid, PQExpBufferData *inoutbuf, PQExpBuffe
 			char *col;
 			char *dist_columns = PQgetvalue(result1, 0, 0);
 			char *dist_colname;
-			if(dist_columns && strlen(dist_columns) > 0)
+			if (dist_columns && !strcmp(dist_columns, "{-128}"))
```
Comment (bhuvnesh2703) [inline src/bin/psql/describe.c]: instead of -128, please use GpReplicatedTableAttributerNumberIdentifier.

File: src/backend/catalog/heap.c [hunk 11]
```diff
@@ -1881,16 +1881,21 @@ heap_create_with_catalog(const char *relname,
...
-        new_rel_desc->rd_cdbpolicy = GpPolicyCopy(GetMemoryChunkContext(new_rel_desc), policy);
-        GpPolicyStore(relid, policy);
-    }
+	 * CDB: If caller gave us a distribution policy, store the distribution
+	 * key column list in the gp_distribution_policy catalog and attach a
+	 * copy to the relcache entry.
+	 *
+	 * Notice: we only store policy for replicated table on segment now.
+	 */
+	if (policy &&
```
Comment (bhuvnesh2703) [inline src/backend/catalog/heap.c]: should we add some checks in gpcheckcat (catalog checking utility) to identify any inconsistencies as we are now having the distribution policy for such tables on each segment?

Comment (Eulerizeit) [reply src/backend/catalog/heap.c]: I think this feature may provide some impetus to refactor gpcheckcat.

File: src/backend/commands/copy.c [hunk 12]
```diff
@@ -1554,6 +1554,16 @@ DoCopyInternal(const CopyStmt *stmt, const char *queryString, CopyState cstate)
 					 errmsg("table \"%s\" does not have OIDs",
 							RelationGetRelationName(cstate->rel))));
 
+#if 0
```
Comment (bhuvnesh2703) [inline src/backend/commands/copy.c]: why if false? just checked the copy feature works..any issues due to which you plan to disable it?

File: src/backend/commands/copy.c [hunk 13]
```diff
@@ -1554,6 +1554,16 @@ DoCopyInternal(const CopyStmt *stmt, const char *queryString, CopyState cstate)
...
+		/* FIXME: Don't allow COPY to from a replicated table */
+		if (GpPolicyIsReplicated(cstate->rel->rd_cdbpolicy))
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("COPY to/from a replicated table is not support")));
```
Comment (bhuvnesh2703) [inline src/backend/commands/copy.c]: `support` ==> `supported`

File: src/backend/cdb/cdbcat.c [hunk 14]
```diff
@@ -209,13 +215,20 @@ GpPolicyFetch(MemoryContext mcxt, Oid tbloid)
...
+		/* covert policy type to replicated if needed */
```
Comment (bhuvnesh2703) [inline src/backend/cdb/cdbcat.c]: `covert` ==> `convert`

File: src/backend/cdb/cdbcopy.c [hunk 15]
```diff
@@ -337,6 +338,13 @@ cdbCopyGetData(CdbCopy *c, bool copy_cancel, uint64 *rows_processed)
 			 */
 			if (nbytes > 0 && buffer)
 			{
+				/* FIXME: only accept data from seg0 for replicated table */
```
Comment (bhuvnesh2703) [inline src/backend/cdb/cdbcopy.c]: fixme needs to be removed?

**Comment by pengzhout** [comment]
@hlinnaka  I'm considering your suggestion of putting replicated table to both QD and QEs, two concerns are:
1. this strategy means we need a new group of workers to manipulate the replicated table, it may be a new N+1 gang or a mix of existed N-gang and 1-gang, I tried to add such gang before (to resolve the problems that pg_locks within CTAS and INSERT only produce partial results) and it's even a bigger change (bunch of dispatcher, interconnect changes).

2. this means we need to store non-catalog table on master which is not a convention in GPDB and it may put master under more stress.

So my thought is we put replicated table only on segments for now, then I can revise my previous commit about N+1 gangs, I think it's also a very useful feature, then if everyone have no concern about putting non-catalog table data on master, I can change it back quickly. What do you think?

**Comment by pengzhout** [comment]
@bhuvnesh2703 thanks a lot, will fix them, can I pick your fix about orca to this PR?

**Comment by pengzhout** [comment]
rebased with the latest code of master and resolved conflicts.

File: src/backend/parser/parse_utilcmd.c (hunk 8 above)
Comment (pengzhout) [reply src/backend/parser/parse_utilcmd.c]: fixed

File: src/backend/parser/parse_utilcmd.c (hunk 9 above)
Comment (pengzhout) [reply src/backend/parser/parse_utilcmd.c]: fixed, thanks

File: src/bin/psql/describe.c (hunk 10 above)
Comment (pengzhout) [reply src/bin/psql/describe.c]: fixed, already abandon such hack way

File: src/backend/catalog/heap.c (hunk 11 above)
Comment (pengzhout) [reply src/backend/catalog/heap.c]: we are definitely going to consider utilities like gpcheckcat/pgdump/gpexpand, but they are not the scope of this PR, we will do that later.

**Comment by pengzhout** [comment]
@vraghavan78 @bhuvnesh2703 I created a PR to fallback queries over replicated table, please help to review, thanks. https://github.com/greenplum-db/gporca/pull/352

**Comment by pengzhout** [comment]
@danielgustafsson Checked that exchange only works for relations have the same distribution policy, so exchange between partitioned table and replicated table will report an error.

**Comment by pengzhout** [comment]
pushed a commit to support gpexpand.

**Comment by pengzhout** [comment]
Merged, thanks for the review.

**Comment by NB13** [comment]
@pengzhout thank you for your idea and implementation! Do you know when this feature will be released as part of new GPDB version?
